diff --git a/BUILD.gn b/BUILD.gn
index ac37521a7..110bb5d20 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -98,6 +98,7 @@ config("extra_warnings") {
   if (is_clang) {
     # Remove when crbug.com/428099 is resolved.
     cflags += [ "-Winconsistent-missing-override" ]
+    cflags += [ "-Wno-implicit-fallthrough" ]
   }
 }
 
diff --git a/src/vulkan_support/BUILD.gn b/src/vulkan_support/BUILD.gn
index 86376012d..668dac570 100644
--- a/src/vulkan_support/BUILD.gn
+++ b/src/vulkan_support/BUILD.gn
@@ -244,6 +244,7 @@ config("vulkan_internal_config") {
     # See https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/1649
     # TODO(jmadill): Lift this once github issue is fixed.
     cflags += [ "-Wno-ignored-attributes" ]
+    cflags += [ "-Wno-implicit-fallthrough" ]
   }
   if (is_linux) {
     defines += [
@@ -475,6 +476,11 @@ config("spirv_tools_config") {
     ]
   }
 }
+config("spirv_tools_internal_config") {
+  if (is_clang) {
+    cflags = [ "-Wno-implicit-fallthrough" ]
+  }
+}
 
 static_library("spirv_tools") {
   deps = [
@@ -565,6 +571,7 @@ static_library("spirv_tools") {
     "$spirv_tools_dir/source/validate_type_unique.cpp",
   ]
   public_configs = [ ":spirv_tools_config" ]
+  configs += [ ":spirv_tools_internal_config" ]
 
   foreach(target_name, spvtools_vendor_tables) {
     deps += [ ":$target_name" ]
@@ -588,6 +595,9 @@ config("glslang_internal_config") {
       "-Wno-reorder",
     ]
   }
+  if (is_clang) {
+    cflags += [ "-Wno-implicit-fallthrough" ]
+  }
 }
 
 static_library("glslang") {
