diff --git a/gni/v8.gni b/gni/v8.gni
index ce9953ac68..e3c5339c2e 100644
--- a/gni/v8.gni
+++ b/gni/v8.gni
@@ -72,7 +72,9 @@ v8_inspector_js_protocol = v8_path_prefix + "/src/inspector/js_protocol.json"
 #
 
 # Common configs to remove or add in all v8 targets.
-v8_remove_configs = []
+v8_remove_configs = [
+  "//build/config/compiler:warn_fallthrough"
+]
 v8_add_configs = [
   v8_path_prefix + ":features",
   v8_path_prefix + ":toolchain",
