diff --git a/third_party/protobuf/src/google/protobuf/descriptor_database_unittest.cc b/third_party/protobuf/src/google/protobuf/descriptor_database_unittest.cc
index 7e81d70af236..3e93b60fc615 100644
--- a/third_party/protobuf/src/google/protobuf/descriptor_database_unittest.cc
+++ b/third_party/protobuf/src/google/protobuf/descriptor_database_unittest.cc
@@ -73,8 +73,6 @@ static void ExpectContainsType(const FileDescriptorProto& proto,

 // ===================================================================

-#if GTEST_HAS_PARAM_TEST
-
 // SimpleDescriptorDatabase, EncodedDescriptorDatabase, and
 // DescriptorPoolDatabase call for very similar tests.  Instead of writing
 // three nearly-identical sets of tests, we use parameterized tests to apply
@@ -482,8 +480,6 @@ INSTANTIATE_TEST_CASE_P(MemoryConserving, DescriptorDatabaseTest,
 INSTANTIATE_TEST_CASE_P(Pool, DescriptorDatabaseTest,
     testing::Values(&DescriptorPoolDatabaseTestCase::New));

-#endif  // GTEST_HAS_PARAM_TEST
-
 TEST(EncodedDescriptorDatabaseExtraTest, FindNameOfFileContainingSymbol) {
   // Create two files, one of which is in two parts.
   FileDescriptorProto file1, file2a, file2b;
